### ISSUE TITLE
Function App instance with a single Durable Entity in it produced tons of 'EventRaised' records in its *History table

### ISSUE THREAD

**Issue by scale-tone** [Azure/azure-functions-durable-extension#1826]
My test instance of [this code](https://github.com/scale-tone/sftp-watcher) went into a weird state.
There's one single `SftpToStorageQueueWatcherEntity` entity in it, that entity is periodically signalled by a [TimerTrigger](https://github.com/scale-tone/sftp-watcher/blob/main/TimerTrigger.cs).
According to logs, the TimerTrigger always succeeds (aka successfully sends a signal to the entity). But the entity doesn't get that signal, according to logs. And its state isn't updated.

And the underlying *History table now contains > 300K records that look like this:

> @sftptostoragequeuewatcherentity@bfkrmsiwbkubo.northeurope.azurecontainer.io-*.txt,0000000000050088,2021-05-06T22:08:09.076Z,-1,Edm.Int32,EventRaised,Edm.String,1552010a9ddc4f02954f3b3be4802126,Edm.String,false,Edm.Boolean,2021-05-06T20:47:25.005Z,Edm.DateTime,"{""op"":""Watch"",""signal"":true,""input"":""\""bfkrmsiwbkubo.northeurope.azurecontainer.io/*.txt\"""",""id"":""0226d4b7-9ec2-41e1-aab9-9bea19b70326"",""Timestamp"":""0001-01-01T00:00:00"",""Predecessor"":""0001-01-01T00:00:00""}",Edm.String,op,Edm.String,,,,

, which keep coming.

I can also see lots of weird logs like this in the AppInsights 'traces' table:

> @sftptostoragequeuewatcherentity@bfkrmsiwbkubo.northeurope.azurecontainer.io-*.txt: Function 'sftptostoragequeuewatcherentity (Entity)' completed 'Watch' operation 8a49fc8c-ef15-42ec-a598-5350ed7fab0a in 2060.9566ms. IsReplay: False. Input: (1612 bytes). Output: (null). HubName: sftpw2mkeaxozmpyvmfunction. AppName: sftpw2mkeaxozmpyvm-function. SlotName: Production. ExtensionVersion: 2.4.0. SequenceNumber: 393.
> 
> @sftptostoragequeuewatcherentity@bfkrmsiwbkubo.northeurope.azurecontainer.io-*.txt: Function 'sftptostoragequeuewatcherentity (Entity)' completed 'Watch' operation 427377a5-a6db-4ccc-b23a-b6ac21e094d9 in 1512.2289ms. IsReplay: False. Input: (1612 bytes). Output: (null). HubName: sftpw2mkeaxozmpyvmfunction. AppName: sftpw2mkeaxozmpyvm-function. SlotName: Production. ExtensionVersion: 2.4.0. SequenceNumber: 394.
> 
> @sftptostoragequeuewatcherentity@bfkrmsiwbkubo.northeurope.azurecontainer.io-*.txt: Function 'sftptostoragequeuewatcherentity (Entity)' completed 'Watch' operation f1951c7b-bfa6-4ec4-9841-8aefb8c1798a in 1407.3627ms. IsReplay: False. Input: (1612 bytes). Output: (null). HubName: sftpw2mkeaxozmpyvmfunction. AppName: sftpw2mkeaxozmpyvm-function. SlotName: Production. ExtensionVersion: 2.4.0. SequenceNumber: 395.
> 

I have another, healthy instance running (separate Storage) - and it doesn't do anything like that (its *History table is stable and only contains 8 records, the entity works as expected and produces expected logs).

Why is all of this and how do I prevent this from happening?

### Expected behavior

The entity successfully receives its signals and processes them.

### Actual behavior

The entity doesn't get invoked, and the *History table constantly inflates in size.

### Relevant source code snippets

https://github.com/scale-tone/sftp-watcher

### App Details

- **Durable Functions extension version**: 2.4.0
- **Azure Functions runtime version**: 3.0
- **Programming language used**: C#

### Screenshots

![image](https://user-images.githubusercontent.com/5447190/117375967-dc78b800-aed8-11eb-9ff3-f02ae99006a6.png)

![image](https://user-images.githubusercontent.com/5447190/117376513-f49d0700-aed9-11eb-9b10-7f385ded40dd.png)

### If deployed to Azure

- **Timeframe issue observed**: still observed
- **Function App name**: sftpw2mkeaxozmpyvm-function
- **Azure region**: North Europe
- **Orchestration instance ID(s)**: @sftptostoragequeuewatcherentity@bfkrmsiwbkubo.northeurope.azurecontainer.io-*.txt
- **Azure storage account name**: sftpw2mkeaxozmpyvmst

**Comment by olitomlinson** [comment]
@scale-tone

I've seen entity signals constantly retry (like in your example) when they couldn't be written to storage due to illegal characters in the entity `instance id`.

Long shot, but can you compare the `instance Id` of the entity which is causing problems, to an `instance Id` which isn't causing problems? Paying attention to the kind of characters used in the Id itself.

I ask because I notice that the `instance Id` of your problematic entity consists of a few non-alphanumeric characters. `bfkrmsiwbkubo.northeurope.azurecontainer.io-*.txt` such as `-`, `*`, and `.`

I believe these characters are okay for table storage, hence why you can see them successfully written to the history table as `Partition key`, but I do wonder if there is some other weirdness going on elsewhere that might be causing a problem with these characters.

---

Another thought...

Entity operations will repeat forever until successful. So is there anything in the DI aspects of bringing up this particular Entity that is different to other Entities? Is there a difference in the runtime code-path for this particular entity instance that is different to other instances, maybe in the constructor?

**Comment by cgillum** [comment]
That’s very odd. I can’t think of a reason why we’d write to the history without first invoking the entity, unless maybe there are certain failure that the entity dispatcher is swallowing. We’ve seen some strange behaviors, for example, when DI code fails in the constructor, but you’re saying there are no errors at all?

At this point if all those 300K+ history events are for the same entity, we probably don’t have enough memory to load them anymore.

@sebastianburckhardt any ideas?

**Comment by sebastianburckhardt** [comment]
There are indeed a few situations in which an entity can receive messages and update the history table without executing operations. For example, if operation messages are received out of order, a message arriving early gets buffered before being delivered. Also, lock messages do not cause operation executions.

This is likely a bug. The history of entities should never grow! It always calls continue-as-new after processing events.

I will need to take a look. I am suspecting the MessageSorter has some error with how DateTime is serialized/deserialized for different locales, which could perhaps also explain the behavioral differences between different deployments.

If this is indeed the problem, then as a quick workaround, you can try to disable the message sorter in host.json
```
{
   "extensions": {
      ... other settings omitted for brevity ...
      "durableTask": {
          "EntityMessageReorderWindow" : "0"
      }
   }
}
```
Turning the message sorter off means we do not deduplicate and sort the entity signals, which can be an issue for some applications but it may be o.k. for your use case.

**Comment by scale-tone** [comment]
> but you’re saying there are no errors at all?

Actually, I managed to catch an error (it wasn't obvious, because this error only shows up two times per hour):

```
2021-05-21T19:32:41.979 [Error] Timeout value of 00:30:00 exceeded by function 'SftpToStorageQueueWatcherEntity' (Id: '644abd2a-fa25-48e3-b49e-ca25c23b51fe'). Initiating cancellation.
2021-05-21T19:32:42.088 [Error] Executed 'SftpToStorageQueueWatcherEntity' (Failed, Id=644abd2a-fa25-48e3-b49e-ca25c23b51fe, Duration=1800149ms)Timeout value of 00:30:00 was exceeded by function: SftpToStorageQueueWatcherEntity
```

And I managed to find the same errors in traces:
![image](https://user-images.githubusercontent.com/5447190/119192367-b894a500-ba88-11eb-92dd-f8aecac224cd.png)

(they only started from 06.05 because App Insights were not configured before that date)

First four lines there:

```
<!--StartFragment-->
timestamp [UTC]2021-05-06T22:27:00.1825594Z messageBreaking change analysis operation failed severityLevel2 itemTypetracecustomDimensions{"prop__{OriginalFormat}":"Breaking change analysis operation failed","ProcessId":"12792","HostInstanceId":"689950c1-401c-4316-963c-d7211e76fa4a","Category":"Microsoft.Azure.WebJobs.Script.ChangeAnalysis.ChangeAnalysisService","LogLevel":"Warning"} client_TypePC client_IP0.0.0.0 cloud_RoleNamesftpw2mkeaxozmpyvm-function cloud_RoleInstance3cdef757a14280590f75eb380fa0efe3ecb8d02e68be4ae537050273878f708c appId650651d0-8243-4d8a-9e66-489b829b5fe7 appNamesftpw2mkeaxozmpyvm-function-insights iKey48c0ee13-0732-4b89-b270-6448855ff525 sdkVersionazurefunctions: 3.0.15584.0 itemId3e938f2a-aeba-11eb-ac52-8be244255f1f itemCount1 | &nbsp; | timestamp [UTC] | 2021-05-06T22:27:00.1825594Z | &nbsp; | message | Breaking change analysis operation failed | &nbsp; | severityLevel | 2 | &nbsp; | itemType | trace | &nbsp; | customDimensions | {"prop__{OriginalFormat}":"Breaking change analysis operation failed","ProcessId":"12792","HostInstanceId":"689950c1-401c-4316-963c-d7211e76fa4a","Category":"Microsoft.Azure.WebJobs.Script.ChangeAnalysis.ChangeAnalysisService","LogLevel":"Warning"} | &nbsp; | client_Type | PC | &nbsp; | client_IP | 0.0.0.0 | &nbsp; | cloud_RoleName | sftpw2mkeaxozmpyvm-function | &nbsp; | cloud_RoleInstance | 3cdef757a14280590f75eb380fa0efe3ecb8d02e68be4ae537050273878f708c | &nbsp; | appId | 650651d0-8243-4d8a-9e66-489b829b5fe7 | &nbsp; | appName | sftpw2mkeaxozmpyvm-function-insights | &nbsp; | iKey | 48c0ee13-0732-4b89-b270-6448855ff525 | &nbsp; | sdkVersion | azurefunctions: 3.0.15584.0 | &nbsp; | itemId | 3e938f2a-aeba-11eb-ac52-8be244255f1f | &nbsp; | itemCount | 1
-- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | --
&nbsp; | timestamp [UTC] | 2021-05-06T22:27:00.1825594Z
&nbsp; | message | Breaking change analysis operation failed
&nbsp; | severityLevel | 2
&nbsp; | itemType | trace
&nbsp; | customDimensions | {"prop__{OriginalFormat}":"Breaking change analysis operation failed","ProcessId":"12792","HostInstanceId":"689950c1-401c-4316-963c-d7211e76fa4a","Category":"Microsoft.Azure.WebJobs.Script.ChangeAnalysis.ChangeAnalysisService","LogLevel":"Warning"}
&nbsp; | client_Type | PC
&nbsp; | client_IP | 0.0.0.0
&nbsp; | cloud_RoleName | sftpw2mkeaxozmpyvm-function
&nbsp; | cloud_RoleInstance | 3cdef757a14280590f75eb380fa0efe3ecb8d02e68be4ae537050273878f708c
&nbsp; | appId | 650651d0-8243-4d8a-9e66-489b829b5fe7
&nbsp; | appName | sftpw2mkeaxozmpyvm-function-insights
&nbsp; | iKey | 48c0ee13-0732-4b89-b270-6448855ff525
&nbsp; | sdkVersion | azurefunctions: 3.0.15584.0
&nbsp; | itemId | 3e938f2a-aeba-11eb-ac52-8be244255f1f
&nbsp; | itemCount | 1

<!--EndFragment-->
```

```
<!--StartFragment-->
timestamp [UTC] | 2021-05-06T22:56:43.9543218Z
-- | --
&nbsp; | message | Timeout value of 00:30:00 exceeded by function 'SftpToStorageQueueWatcherEntity' (Id: '5a528d16-4f2f-4b3e-a296-9213647e24fc'). Initiating cancellation.
&nbsp; | severityLevel | 3
&nbsp; | itemType | trace
&nbsp; | customDimensions | {"prop__{OriginalFormat}":"Timeout value of 00:30:00 exceeded by function 'SftpToStorageQueueWatcherEntity' (Id: '5a528d16-4f2f-4b3e-a296-9213647e24fc'). Initiating cancellation.","ProcessId":"12792","HostInstanceId":"689950c1-401c-4316-963c-d7211e76fa4a","Category":"Function.SftpToStorageQueueWatcherEntity","LogLevel":"Error","InvocationId":"5a528d16-4f2f-4b3e-a296-9213647e24fc"}
&nbsp; | operation_Name | SftpToStorageQueueWatcherEntity
&nbsp; | operation_Id | c1c49a620a922b438cb5c80d313acf23
&nbsp; | operation_ParentId | c2883bc35554ed4c
&nbsp; | client_Type | PC
&nbsp; | client_IP | 0.0.0.0
&nbsp; | cloud_RoleName | sftpw2mkeaxozmpyvm-function
&nbsp; | cloud_RoleInstance | 3cdef757a14280590f75eb380fa0efe3ecb8d02e68be4ae537050273878f708c
&nbsp; | appId | 650651d0-8243-4d8a-9e66-489b829b5fe7
&nbsp; | appName | sftpw2mkeaxozmpyvm-function-insights
&nbsp; | iKey | 48c0ee13-0732-4b89-b270-6448855ff525
&nbsp; | sdkVersion | azurefunctions: 3.0.15584.0
&nbsp; | itemId | 54232228-aebe-11eb-ac52-8be244255f1f
&nbsp; | itemCount | 1

<!--EndFragment-->
```

```
<!--StartFragment-->
timestamp [UTC] | 2021-05-06T22:56:44.005678Z
-- | --
&nbsp; | message | Executed 'SftpToStorageQueueWatcherEntity' (Failed, Id=5a528d16-4f2f-4b3e-a296-9213647e24fc, Duration=1800143ms)
&nbsp; | severityLevel | 3
&nbsp; | itemType | trace
&nbsp; | customDimensions | {"prop__{OriginalFormat}":"Executed '{functionName}' ({status}, Id={invocationId}, Duration={executionDuration}ms)","ProcessId":"12792","HostInstanceId":"689950c1-401c-4316-963c-d7211e76fa4a","Category":"Function.SftpToStorageQueueWatcherEntity","prop__functionName":"SftpToStorageQueueWatcherEntity","LogLevel":"Error","InvocationId":"5a528d16-4f2f-4b3e-a296-9213647e24fc","EventName":"FunctionCompleted","EventId":"3","prop__invocationId":"5a528d16-4f2f-4b3e-a296-9213647e24fc","prop__executionDuration":"1800143","prop__status":"Failed"}
&nbsp; | operation_Name | SftpToStorageQueueWatcherEntity
&nbsp; | operation_Id | c1c49a620a922b438cb5c80d313acf23
&nbsp; | operation_ParentId | c2883bc35554ed4c
&nbsp; | client_Type | PC
&nbsp; | client_IP | 0.0.0.0
&nbsp; | cloud_RoleName | sftpw2mkeaxozmpyvm-function
&nbsp; | cloud_RoleInstance | 3cdef757a14280590f75eb380fa0efe3ecb8d02e68be4ae537050273878f708c
&nbsp; | appId | 650651d0-8243-4d8a-9e66-489b829b5fe7
&nbsp; | appName | sftpw2mkeaxozmpyvm-function-insights
&nbsp; | iKey | 48c0ee13-0732-4b89-b270-6448855ff525
&nbsp; | sdkVersion | azurefunctions: 3.0.15584.0
&nbsp; | itemId | 54232229-aebe-11eb-ac52-8be244255f1f
&nbsp; | itemCount | 1

<!--EndFragment-->
```

```
<!--StartFragment-->
timestamp [UTC]2021-05-06T22:56:44.0356639Z messageA function timeout has occurred. Host is shutting down. severityLevel3 itemTypetracecustomDimensions{"prop__{OriginalFormat}":"A function timeout has occurred. Host is shutting down.","ProcessId":"12792","HostInstanceId":"689950c1-401c-4316-963c-d7211e76fa4a","Category":"Microsoft.Azure.WebJobs.Script.WebHost.WebScriptHostExceptionHandler","LogLevel":"Error"} client_TypePC client_IP0.0.0.0 cloud_RoleNamesftpw2mkeaxozmpyvm-function cloud_RoleInstance3cdef757a14280590f75eb380fa0efe3ecb8d02e68be4ae537050273878f708c appId650651d0-8243-4d8a-9e66-489b829b5fe7 appNamesftpw2mkeaxozmpyvm-function-insights iKey48c0ee13-0732-4b89-b270-6448855ff525 sdkVersionazurefunctions: 3.0.15584.0 itemId5423222a-aebe-11eb-ac52-8be244255f1f itemCount1 | &nbsp; | timestamp [UTC] | 2021-05-06T22:56:44.0356639Z | &nbsp; | message | A function timeout has occurred. Host is shutting down. | &nbsp; | severityLevel | 3 | &nbsp; | itemType | trace | &nbsp; | customDimensions | {"prop__{OriginalFormat}":"A function timeout has occurred. Host is shutting down.","ProcessId":"12792","HostInstanceId":"689950c1-401c-4316-963c-d7211e76fa4a","Category":"Microsoft.Azure.WebJobs.Script.WebHost.WebScriptHostExceptionHandler","LogLevel":"Error"} | &nbsp; | client_Type | PC | &nbsp; | client_IP | 0.0.0.0 | &nbsp; | cloud_RoleName | sftpw2mkeaxozmpyvm-function | &nbsp; | cloud_RoleInstance | 3cdef757a14280590f75eb380fa0efe3ecb8d02e68be4ae537050273878f708c | &nbsp; | appId | 650651d0-8243-4d8a-9e66-489b829b5fe7 | &nbsp; | appName | sftpw2mkeaxozmpyvm-function-insights | &nbsp; | iKey | 48c0ee13-0732-4b89-b270-6448855ff525 | &nbsp; | sdkVersion | azurefunctions: 3.0.15584.0 | &nbsp; | itemId | 5423222a-aebe-11eb-ac52-8be244255f1f | &nbsp; | itemCount | 1
-- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | --
&nbsp; | timestamp [UTC] | 2021-05-06T22:56:44.0356639Z
&nbsp; | message | A function timeout has occurred. Host is shutting down.
&nbsp; | severityLevel | 3
&nbsp; | itemType | trace
&nbsp; | customDimensions | {"prop__{OriginalFormat}":"A function timeout has occurred. Host is shutting down.","ProcessId":"12792","HostInstanceId":"689950c1-401c-4316-963c-d7211e76fa4a","Category":"Microsoft.Azure.WebJobs.Script.WebHost.WebScriptHostExceptionHandler","LogLevel":"Error"}
&nbsp; | client_Type | PC
&nbsp; | client_IP | 0.0.0.0
&nbsp; | cloud_RoleName | sftpw2mkeaxozmpyvm-function
&nbsp; | cloud_RoleInstance | 3cdef757a14280590f75eb380fa0efe3ecb8d02e68be4ae537050273878f708c
&nbsp; | appId | 650651d0-8243-4d8a-9e66-489b829b5fe7
&nbsp; | appName | sftpw2mkeaxozmpyvm-function-insights
&nbsp; | iKey | 48c0ee13-0732-4b89-b270-6448855ff525
&nbsp; | sdkVersion | azurefunctions: 3.0.15584.0
&nbsp; | itemId | 5423222a-aebe-11eb-ac52-8be244255f1f
&nbsp; | itemCount | 1

<!--EndFragment-->
```


Now it indeed seems like the runtime tries to execute a signal by first loading everything from History table, that operation takes forever, and that's the end of it.
Question is still how did we even manage to run into this trouble.
As you can see from the code, it does makes network calls, which do have a chance to occasionally fail/timeout, but it shouldn't lead to such catastrophic consequences, right?


> At this point if all those 300K+ history events are for the same entity, we probably don’t have enough memory to load them anymore.

Yes, that's how I noticed that: DfMon once failed to load this entity's status.


>  as a quick workaround, you can try to disable the message sorter in host.json

@sebastianburckhardt , I assume that won't bring this particular instance back to life (in fact, I already accepted this loss).
But would you say that it can prevent similar thing to happen in future?

**Comment by scale-tone** [comment]
Another bad news is that my another instance (separate storage), which once was healthy, is now unhealthy as well. That started from 15.05, and I'm sure I didn't touch it.
Symptoms are the same. 
Sadly no logs there, because AppInsights are also not configured.

**Comment by scale-tone** [comment]
@cgillum , suppose entity signals are handled slower than are being sent. How soon would it lead us to a disaster like the above?

**Comment by sebastianburckhardt** [comment]
> Question is still how did we even manage to run into this trouble.

Indeed, that is the question. It must be a bug. There is no valid scenario in which the entity history grows. Even if an entity is overloaded, we would just expect control queues to get backed up.

**Comment by sebastianburckhardt** [comment]
I have done a bit of investigation. Although I cannot get traces all the way back to the root cause (the first time the instance history is fetched is 05/06/2021 20:04:30 at which point it already contains a ridiculous 327278 events), I can see a pattern where the Function that processes the entity operations times out in the middleware (before reaching 'FunctionStarted'). It appears we don't handle this kind of timeout correctly.  I suppose I was expecting the timeout to somehow abandon the work item. But instead, apparently, DTFx still commits the events to the history, even though the function was never run. So this process then just repeats, with the history growing and growing. 

I will need to think more about the best way to fix this and handle that situation.

**Comment by scale-tone** [comment]
@sebastianburckhardt , can you think of any workarounds?

**Comment by scale-tone** [comment]
Did it still not make it into v2.6.0? :(

**Comment by sebastianburckhardt** [comment]
Apologies. PR #1917 has proven to drag on much longer than originally expected, as improving the timeout handling requires subtle changes, and because we needed to add tests, which then found more problems (e.g. for out-of-proc timeouts). Still working on it now.

**Comment by cgillum** [comment]
This fix for this was released as part of [v2.6.1](https://github.com/Azure/azure-functions-durable-extension/releases/tag/v2.6.1)
